synapse/rest/client/receipts.py

```diff
@@ -109,6 +109,46 @@ async def on_POST(
 
         return 200, {}
 
+    async def _is_valid_thread_id(self, event_id: str, thread_id: str) -> bool:
+        """
+        The thread ID provided must relate (in a vague sense) to the event ID.
+
+        We check this to ensure clients aren't sending bogus receipts.
+
+        A thread ID is considered valid if:
+
+        1. The event has a thread relation which matches the thread ID.
+        2. The event has children events which form a thread relation (i.e. the
```

Comment:
Some nitpicks here.

- Does "children events" mean direct children or arbitrary descendants?
- Why "events" plural here? Presumably the existence of at least one such child events is sufficient?
- "form a thread relation" presumably means that they are marked as relating (via m.thread) _to the event in question_.

Reply:
> Does "children events" mean direct children or arbitrary descendants?

A thread relation cannot relate to an event which has a relation, so it means direct child.

> Why "events" plural here? Presumably the existence of at least one such child events is sufficient?

Sloppiness, one child is sufficient.

> "form a thread relation" presumably means that they are marked as relating (via m.thread) _to the event in question_.

Yes. It is trying to say that event `A` has another event `B` with an `m.relates_to` of type `m.thread` pointed at `A`.

Reply:
I think we have are working with different meanings of "child". I understand "A is a child of B" to mean "A cites B as a prev_event". But it sounds here that you understand "A is a child of B" to mean "A declares itself as being `m.relates_to` B via `m.thread`". (See also the other comment thread about terminology).

Let me see if I can make a suggestion...

Reply:
I'd phrase it like this, but I do lean towards the mathematical and the formal. Just a suggestion, feel free to ignore!

```
        The event must be related to the thread ID (in a vague sense) to ensure
        clients aren't sending bogus receipts.
        
        A thread ID is considered valid for a given event E if:
        
        1. E has a thread relation which matches the thread ID;
        2. some other event has a thread relation to E matching the thread ID; or
        3. E is recursively related (via any rel_types) to an event
           which satisfies 1 or 2.
```